### ISSUE TITLE
Keep getting messages with empty value along with same messages repeatedly

### ISSUE THREAD

**Issue by 0xmtn** [SOHU-Co/kafka-node#1059]
Hello,

I'm trying to use kafka, kafka-node and lua-mosquitto together. Assume we have a structure like below:

client1:
    - Subscribes to: _c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_k
    - Publishes to: _c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e

client2:
    - Subscribes to: _c_5b8b215c855b7569eb94f0ed_d_ 5b8b26dd2ac7176a44765c0c_k
    - Publishes to: _c_5b8b215c855b7569eb94f0ed_d_ 5b8b26dd2ac7176a44765c0c_e
 
Center:
   - Subscribes to: every topic that clients publishe to.
   - Publishes to: every topic that clients subscribe to, at will. 

Now when I publish something from client1 to topic _c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_k, I assume that Center would get it. When the Center is stopped and client1 publishes, and Center is started, messages start flood into Center Consumer. I've only tested one client and one topic. 

**_Most of the messages has value='', and offset=0. How can I not get these messages?_**

### Environment
* Node version: v10.9.0
* Kafka-node version: 2.6.1
* Kafka version: 2.11-2.0.0

###Setup
* Number of Brokers: 3
* Number partitions for topic: 1

```js
var kafka_client = new kafka.Client("localhost:2181", "KW");

kafka_client.zk.client.getChildren("/brokers/topics", (err, children, stats) => {
    children.forEach(function(child){
      if(child.slice(-1) === "e") { // Get the topic with char `e` at the end 
        var KWConsumer = new Consumer(kafka_client, [{topic: child}]);
        KWConsumer.on("message", function(message){
          console.log('%s read msg Topic="%s" Partition=%s Offset=%d Value=%s' , this.client.clientId, message.topic, message.partition, message.offset, message.value);
          /*KWConsumer.commit(function(err, data){
            console.log("Committed: ", err, data);
          });*/
        });
      } 
    }); 
});
```

```lua
mqtt = require("mosquitto")
client = mqtt.new()

client.ON_CONNECT = function()
    print("5b8bc0aa6be58c05415cbc44 connected")
    client:publish("_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e", "hello")
    client:subscribe("_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_k")
end

client.ON_MESSAGE = function(mid, topic, payload)
    print(mid, topic, payload)
end

client.ON_PUBLISH = function()
    
    -- client:disconnect()
end

client:connect() #Connects to Mosca server with kafka backend - localhost:1883
client:loop_forever()
```

Lua MQtt Clients connect to [https://github.com/mcollina/mosca](Mosca) server with kafka backend.
```Mosca Server 
var backend = {
    type: "kafka",
    json: false,
    connectionString: "localhost:2181",
    clientId: "mosca",
    groupId: "mosca",
    defaultEncoding: "utf8",
  };

var moscaSettings = {
    interfaces: [
        { type: "mqtt", port: 1883 },
    ],
    id: "mosca",
    stats: false,
    publishNewClient: false,
    publishClientDisconnect: false,
    publishSubscriptions: false,
    backend: backend,
};
var server = new mosca.Server(moscaSettings);
server.on('ready', setup);
function setup() {
    console.log('Mosca server is up and running.');
}

server.on("error", function (err) {
    console.log(err);
});
server.on('clientConnected', function (client) {
    console.log('Client Connected \t:= ', client.id);
});
server.on('published', function (packet, client) {
  packet.payload = packet.payload.toString('utf8');
  console.log("Published :=", packet);
});
server.on('subscribed', function (topic, client) {
    console.log("Subscribed :=", topic, client.packet);
});
server.on('unsubscribed', function (topic, client) {
    console.log('unsubscribed := ', topic);
});
server.on('clientDisconnecting', function (client) {
    console.log('clientDisconnecting := ', client.id);
});
server.on('clientDisconnected', function (client) {
    console.log('Client Disconnected     := ', client.id);
});
```

### Output
```
kafka-node:Consumer consumer ready +0ms
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=1 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=2 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=3 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=4 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=5 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=6 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=7 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=8 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=9 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=10 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=11 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=12 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=13 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=14 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=15 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=16 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=17 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=1 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=2 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=3 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=4 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=5 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=6 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=7 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=8 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=9 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=10 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=11 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=12 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=13 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=14 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=15 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=16 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=17 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=1 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=2 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=3 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=4 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=5 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=6 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=7 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=8 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=9 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=10 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=11 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=12 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=13 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=14 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=15 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=16 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=17 Value=hello
KW read msg Topic="_c_5b8b215c855b7569eb94f0ed_d_5b8bc0aa6be58c05415cbc44_e" Partition=0 Offset=0 Value=
```

**NOTE:** When this whole thing happens, Mosca server doesn't output anything. So IMHO it is either kafka-node or Kafka/Zookeeper Server problem. 

Thank you

**Comment by hyperlink** [comment]
I believe this is fixed by #1051 could you check out master and verify?

**Comment by 0xmtn** [comment]
Perfect! Thank you @hyperlink 

Now, I don't get messages with empty value and 0 offset, but I get all of the messages that's been published to the topic before the consumer went live.

I assume the only way to only get messages that haven't been consumed before is to check their offsets against a db? Right? 

Or is there any other, more 'Kafka' way to accomplish this? I.e. Some kind of identifier for the consumer which is saved to Kafka FS and in turn publish only what's not been consumed before? Is it too disneyland?
